File: trunk/managed/Warp3D/Warp3DVB/warp_Scene.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------

Imports System
Imports System.Collections
Imports System.Drawing
Imports System.Timers

Namespace Rednettle.Warp3D
    '/ <summary>
    '/ Summary description for warp_Scene.
    '/ </summary>
    Public Class warp_Scene
        Inherits warp_CoreObject
        Public Shared version As String = "1.0.0"
        Public Shared release As String = "0010"

        Public renderPipeline As warp_RenderPipeline
        Public width As Integer, height As Integer

        Public environment As New warp_Environment '= New warp_Environment()
        Public defaultCamera As warp_Camera = warp_Camera.FRONT()

        Public wobject() As warp_Object
        Public light() As warp_Light

        Public objects As Integer = 0
        Public lights As Integer = 0

        Private objectsNeedRebuild As Boolean = True
        Private lightsNeedRebuild As Boolean = True

        Protected preparedForRendering As Boolean = False

        Public normalizedOffset As warp_Vector = New warp_Vector(0.0F, 0.0F, 0.0F)
        Public normalizedScale As Single = 1.0F
        Private Shared instancesRunning As Boolean = False

        Public objectData As Hashtable = New Hashtable()
        Public lightData As Hashtable = New Hashtable()
        Public materialData As Hashtable = New Hashtable()
        Public cameraData As Hashtable = New Hashtable()

        Dim probes As Integer = 0
        Dim perf As Integer = 0

        Public fps As String = "0"

        Public rendertime As Integer = 500

        Public Sub New()
        End Sub

        Public Sub New(ByVal w As Integer, ByVal h As Integer)
            showInfo()
            width = w
            height = h
            renderPipeline = New warp_RenderPipeline(Me, w, h)
        End Sub

        Public Sub showInfo()
            If instancesRunning Then Exit Sub

            System.Console.WriteLine()
            System.Console.WriteLine("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
            System.Console.WriteLine("WARP 3D Kernel " + version + " [Build " + release + "]")
            System.Console.WriteLine("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")

            instancesRunning = True
        End Sub

        Public Sub removeAllObjects()
            objectData = New Hashtable()
            objectsNeedRebuild = True
            rebuild()
        End Sub

        Public Sub rebuild()

            If objectsNeedRebuild Then

                objectsNeedRebuild = False

                objects = objectData.Count
                wobject = New warp_Object(objects - 1) {}
                Dim enumerator As IDictionaryEnumerator = objectData.GetEnumerator()

                For i As Integer = objects - 1 To 0 Step -1
                    enumerator.MoveNext()
                    wobject(i) = CType(enumerator.Value, warp_Object)

                    wobject(i).id = i
                    wobject(i).rebuild()
                Next
            End If

            If lightsNeedRebuild Then
                lightsNeedRebuild = False
                lights = lightData.Count
                light = New warp_Light(lights - 1) {}
                Dim enumerator As IDictionaryEnumerator = lightData.GetEnumerator()
                For i As Integer = lights - 1 To 0 Step -1
                    enumerator.MoveNext()
                    light(i) = CType(enumerator.Value, warp_Light)

                Next
            End If
        End Sub

        Public Function sceneobject(ByVal key As String) As warp_Object
            Return CType(objectData(key), warp_Object)
        End Function

        Public Function material(ByVal key As String) As warp_Material
            Return CType(materialData(key), warp_Material)
        End Function

        Public Function camera(ByVal key As String) As warp_Camera
            Return CType(cameraData(key), warp_Camera)
        End Function

        Public Sub addObject(ByVal key As String, ByVal obj As warp_Object)
            obj.name = key
            objectData.Add(key, obj)
            obj.parent = Me
            objectsNeedRebuild = True
        End Sub

        Public Sub removeObject(ByVal key As String)
            objectData.Remove(key)
            objectsNeedRebuild = True
            preparedForRendering = False
        End Sub

        Public Sub addMaterial(ByVal key As String, ByVal m As warp_Material)
            materialData.Add(key, m)
        End Sub

        Public Sub removeMaterial(ByVal key As String)
            materialData.Remove(key)
        End Sub

        Public Sub addCamera(ByVal key As String, ByVal c As warp_Camera)
            cameraData.Add(key, c)
        End Sub

        Public Sub removeCamera(ByVal key As String)
            cameraData.Remove(key)
        End Sub

        Public Sub addLight(ByVal key As String, ByVal l As warp_Light)
            lightData.Add(key, l)
            lightsNeedRebuild = True
        End Sub

        Public Sub removeLight(ByVal key As String)
            lightData.Remove(key)
            lightsNeedRebuild = True
            preparedForRendering = False
        End Sub

        Public Sub prepareForRendering()
            If preparedForRendering = True Then
                Return
            End If
            preparedForRendering = True

            System.Console.WriteLine("warp_Scene| prepareForRendering : Preparing for realtime rendering ...")
            rebuild()
            renderPipeline.buildLightMap()
            printSceneInfo()
        End Sub

        Public Sub printSceneInfo()
            System.Console.WriteLine("warp_Scene| Objects   : " & objects)
            System.Console.WriteLine("warp_Scene| Vertices  : " & countVertices())
            System.Console.WriteLine("warp_Scene| Triangles : " & countTriangles())
        End Sub

        Public Sub render()

            Dim n As DateTime = DateTime.Now
            renderPipeline.render(Me.defaultCamera)

            Dim s As TimeSpan = DateTime.Now - n

            perf += s.Milliseconds
            probes += 1
            If probes = 32 Then
                probes = 0
                fps = (perf / 32) & "," & (1000 / (perf / 32))

                perf = 0
            End If

        End Sub

        Public Function getImage() As Bitmap
            Return renderPipeline.screen.getImage()
        End Function

        Public Function getFPS() As String
            Return fps '& " - yes"
        End Function
        ' 		
        ' 		Public java.awt.Dimension size()
        ' 		{
        ' 			Return New java.awt.Dimension(width,height)
        ' 		}
        ' 		*/

        Public Sub setBackgroundColor(ByVal bgcolor As Integer)
            environment.bgcolor = bgcolor
        End Sub

        Public Sub setBackground(ByVal t As warp_Texture)
            environment.setBackground(t)
        End Sub

        Public Sub setAmbient(ByVal ambientcolor As Integer)
            environment.ambient = ambientcolor
        End Sub


        Public Function countVertices() As Integer
            Dim counter As Integer = 0
            For i As Integer = 0 To objects - 1 'Step i + 1
                counter += wobject(i).vertices
            Next
            Return counter
        End Function

        Public Function countTriangles() As Integer
            Dim counter As Integer = 0

            For i As Integer = 0 To objects - 1 'Step i + 1
                counter += wobject(i).triangles
            Next
            Return counter
        End Function

        Public Sub useIdBuffer(ByVal buffer As Boolean)
            renderPipeline.useIdBuffer(Buffer)
        End Sub

        Public Function identifyTriangleAt(ByVal xpos As Integer, ByVal ypos As Integer) As warp_Triangle
            If Not renderPipeline.useId Then Return Nothing
            If xpos < 0 Or xpos >= width Then Return Nothing
            If ypos < 0 Or ypos >= height Then Return Nothing

            Dim pos As Integer = xpos + renderPipeline.screen.width * ypos
            Dim idCode As Integer = renderPipeline.idBuffer(pos)
            If idCode < 0 Then Return Nothing

            Return wobject(idCode >> 16).fasttriangle(idCode And &HFFFF)
        End Function

        Public Function identifyObjectAt(ByVal xpos As Integer, ByVal ypos As Integer) As warp_Object
            Dim tri As warp_Triangle = identifyTriangleAt(xpos, ypos)
            If tri Is Nothing Then Return Nothing
            Return tri.parent
        End Function

        Public Sub normalize()
            objectsNeedRebuild = True
            rebuild()

            Dim min As warp_Vector, max As warp_Vector, tempmax As warp_Vector, tempmin As warp_Vector
            If objects = 0 Then Exit Sub

            matrix = New warp_Matrix()
            normalmatrix = New warp_Matrix()

            max = wobject(0).maximum()
            min = wobject(0).maximum()

            For i As Integer = 0 To objects - 1 'Step i + 1
                tempmax = wobject(i).maximum()
                tempmin = wobject(i).maximum()
                If tempmax.x > max.x Then
                    max.x = tempmax.x
                End If
                If tempmax.y > max.y Then
                    max.y = tempmax.y
                End If
                If tempmax.z > max.z Then
                    max.z = tempmax.z
                End If
                If tempmin.x < min.x Then
                    min.x = tempmin.x
                End If
                If tempmin.y < min.y Then
                    min.y = tempmin.y
                End If
                If tempmin.z < min.z Then
                    min.z = tempmin.z
                End If
            Next
            Dim xdist As Single = max.x - min.x
            Dim ydist As Single = max.y - min.y
            Dim zdist As Single = max.z - min.z
            Dim xmed As Single = (max.x + min.x) / 2
            Dim ymed As Single = (max.y + min.y) / 2
            Dim zmed As Single = (max.z + min.z) / 2

            Dim diameter As Single
            If xdist > ydist Then diameter = xdist Else diameter = ydist
            If zdist > diameter Then diameter = zdist

            normalizedOffset = New warp_Vector(xmed, ymed, zmed)
            normalizedScale = 2 / diameter

            shift(normalizedOffset.reverse())
            scale(normalizedScale)
        End Sub
    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_Scene.cs
using System;
using System.Collections;
using System.Drawing;
using System.Timers;

namespace Warp3D
{
    /// <summary>
    /// Summary description for warp_Scene.
    /// </summary>
    public class warp_Scene : warp_CoreObject
    {
        public static String version = "1.0.0";
        public static String release = "0010";

        public warp_RenderPipeline renderPipeline;
        public int width, height;

        public warp_Environment environment = new warp_Environment();
        public warp_Camera defaultCamera = warp_Camera.FRONT();

        public warp_Object[] wobject;
        public warp_Light[] light;

        public int objects = 0;
        public int lights = 0;

        private bool objectsNeedRebuild = true;
        private bool lightsNeedRebuild = true;

        protected bool preparedForRendering = false;

        public warp_Vector normalizedOffset = new warp_Vector(0f, 0f, 0f);
        public float normalizedScale = 1f;

        public Hashtable objectData = new Hashtable();
        public Hashtable lightData = new Hashtable();
        public Hashtable materialData = new Hashtable();
        public Hashtable cameraData = new Hashtable();

        public warp_Scene()
        {
        }

        public warp_Scene(int w, int h)
        {
            width = w;
            height = h;
            renderPipeline = new warp_RenderPipeline(this, w, h);
            defaultCamera.setScreensize(w, h);
        }

        public void destroy()
        {
            objects = objectData.Count;
            foreach (warp_Object o in objectData.Values)
                o.destroy();

            objectData.Clear();
            lightData.Clear();
            materialData.Clear();
            cameraData.Clear();
            if (renderPipeline != null)
                renderPipeline.Dispose();
            renderPipeline = null;
            environment = null;
            defaultCamera = null;
            wobject = null;
        }

        public void removeAllObjects()
        {

            objectData = new Hashtable();
            objectsNeedRebuild = true;
            rebuild();
        }

        public void rebuild()
        {
            if (objectsNeedRebuild)
            {
                objectsNeedRebuild = false;

                objects = objectData.Count;
                wobject = new warp_Object[objects];
                IDictionaryEnumerator enumerator = objectData.GetEnumerator();

                for (int i = 0; i < objects; ++i)
                {
                    enumerator.MoveNext();
                    wobject[i] = (warp_Object)enumerator.Value;

                    wobject[i].id = i;
                    wobject[i].rebuild();
                }
            }

            if (lightsNeedRebuild)
            {
                lightsNeedRebuild = false;
                lights = lightData.Count;
                light = new warp_Light[lights];
                IDictionaryEnumerator enumerator = lightData.GetEnumerator();
                for (int i = lights - 1; i >= 0; i--)
                {
                    enumerator.MoveNext();
                    light[i] = (warp_Light)enumerator.Value;

                }
            }
        }

        public warp_Object sceneobject(String key)
        {
            return (warp_Object)objectData[key];
        }

        public warp_Material material(String key)
        {
            return (warp_Material)materialData[key];
        }

        public warp_Camera camera(String key)
        {
            return (warp_Camera)cameraData[key];
        }

        public void addObject(String key, warp_Object obj)
        {
            obj.name = key;
            objectData.Add(key, obj);
            obj.parent = this;
            objectsNeedRebuild = true;
            preparedForRendering = false;
        }

        public void removeObject(String key)
        {
            objectData.Remove(key);
            objectsNeedRebuild = true;
            preparedForRendering = false;
        }

        public void addMaterial(String key, warp_Material m)
        {
            materialData.Add(key, m);
        }

        public void removeMaterial(String key)
        {
            materialData.Remove(key);
        }

        public void addCamera(String key, warp_Camera c)
        {
            cameraData.Add(key, c);
        }

        public void removeCamera(String key)
        {
            cameraData.Remove(key);
        }

        public void addLight(String key, warp_Light l)
        {
            lightData.Add(key, l);
            lightsNeedRebuild = true;
        }

        public void removeLight(String key)
        {
            lightData.Remove(key);
            lightsNeedRebuild = true;
            preparedForRendering = false;
        }

        public void prepareForRendering()
        {
            if (preparedForRendering)
                return;
            preparedForRendering = true;

            rebuild();
            renderPipeline.buildLightMap();
        }

        public void render()
        {
            renderPipeline.render(this.defaultCamera);
        }

        public Bitmap getImage()
        {
            return renderPipeline.screen.getImage();
        }

        public void setBackgroundColor(int bgcolor)
        {
            environment.bgcolor = bgcolor | warp_Color.MASKALPHA; // needs to be solid for now
        }

        public void setBackground(warp_Texture t)
        {
            environment.setBackground(t);
        }

        public void setAmbient(int ambientcolor)
        {
            environment.ambient = ambientcolor;
        }

        public int countVertices()
        {
            int counter = 0;
            for (int i = 0; i < objects; i++) counter += wobject[i].vertices;
            return counter;
        }

        public int countTriangles()
        {
            int counter = 0;
            for (int i = 0; i < objects; i++) counter += wobject[i].triangles;
            return counter;
        }

        public void normalize()
        {
            objectsNeedRebuild = true;
            rebuild();

            warp_Vector min, max, tempmax, tempmin;
            if (objects == 0)
            {
                return;
            }

            matrix = new warp_Matrix();
            normalmatrix = new warp_Matrix();

            max = wobject[0].maximum();
            min = wobject[0].maximum();

            for (int i = 0; i < objects; i++)
            {
                tempmax = wobject[i].maximum();
                tempmin = wobject[i].maximum();
                if (tempmax.x > max.x)
                {
                    max.x = tempmax.x;
                }
                if (tempmax.y > max.y)
                {
                    max.y = tempmax.y;
                }
                if (tempmax.z > max.z)
                {
                    max.z = tempmax.z;
                }
                if (tempmin.x < min.x)
                {
                    min.x = tempmin.x;
                }
                if (tempmin.y < min.y)
                {
                    min.y = tempmin.y;
                }
                if (tempmin.z < min.z)
                {
                    min.z = tempmin.z;
                }
            }
            float xdist = max.x - min.x;
            float ydist = max.y - min.y;
            float zdist = max.z - min.z;
            float xmed = (max.x + min.x) / 2;
            float ymed = (max.y + min.y) / 2;
            float zmed = (max.z + min.z) / 2;

            float diameter = (xdist > ydist) ? xdist : ydist;
            diameter = (zdist > diameter) ? zdist : diameter;

            normalizedOffset = new warp_Vector(xmed, ymed, zmed);
            normalizedScale = 2 / diameter;

            shift(normalizedOffset.reverse());
            scale(normalizedScale);
        }

        public float EstimateBoxProjectedArea(warp_Vector pos, warp_Vector size, warp_Matrix rotation)
        {
            warp_Matrix om = new warp_Matrix();
            om.scale(size.x, size.y, size.z);
            om.transform(rotation);
            om.m03 = pos.x;
            om.m13 = pos.y;
            om.m23 = pos.z;

            warp_Matrix m = warp_Matrix.multiply(defaultCamera.getMatrix(), matrix);
            om.transform(m);

            float xmin, xmax;
            float ymin, ymax;
            float zmin;
            warp_Vector side = new warp_Vector(-0.5f, -0.5f, -0.5f);
            warp_Vector v = side.transform(om);
            xmin = v.x;
            xmax = xmin;
            ymin = v.y;
            ymax = ymin;
            zmin = v.z;

            side.x = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = -0.5f;
            side.y = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = -0.5f;
            side.y = -0.5f;
            side.z = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = -0.5f;
            side.y = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            side.x = 0.5f;
            v = side.transform(om);
            if (xmin > v.x)
                xmin = v.x;
            else if (xmax < v.x)
                xmax = v.x;
            if (ymin > v.y)
                ymin = v.y;
            else if (ymax < v.y)
                ymax = v.y;
            if (zmin > v.z)
                zmin = v.z;

            xmax -= xmin;
            ymax -= ymin;

            if (!defaultCamera.isOrthographic)
            {
                float fact = 1.0f / ((zmin > 0.1) ? zmin : 0.1f);
                xmax *= fact;
                ymax *= fact;
            }
            if (xmax < 1f || ymax < 1f)
                return -1;
            return xmax * ymax / (width * height);
        }
    }
}
